streams/dshr_strdata_mod.F90

```diff
@@ -1376,115 +1680,161 @@ subroutine shr_strdata_readstrm(sdat, ns, stream, stream_mesh, &
-    character(len=*)      , intent(in)            :: fldname
-    type(ESMF_Mesh)       , intent(in)            :: stream_mesh
-    type(io_desc_t)       , intent(inout)         :: pio_iodesc
-    integer               , intent(out)           :: rc
+    type(shr_strdata_type)      , intent(in)    :: sdat  ! strdata data data-type
+    type(shr_strdata_perstream) , intent(inout) :: per_stream
+    character(len=*)            , intent(in)    :: fldname
+    type(file_desc_t)           , intent(inout) :: pioid
+    integer                     , intent(out)   :: rc
 
     ! local variables
-    integer                       :: pio_iovartype
-    integer                       :: n
-    type(var_desc_t)              :: varid
-    integer                       :: ndims
-    integer, allocatable          :: dimids(:)
-    integer, allocatable          :: dimlens(:)
-    type(ESMF_DistGrid)           :: distGrid
-    integer                       :: lsize
-    integer, pointer              :: compdof(:) => null()
-    character(CS)                 :: dimname
-    integer                       :: rCode      ! pio return code (only used when pio error handling is PIO_BCAST_ERROR)
-    character(*), parameter       :: subname = '(shr_strdata_set_stream_iodesc) '
-    character(*), parameter       :: F00  = "('(shr_strdata_set_stream_iodesc) ',a,i8,2x,i8,2x,a)"
-    character(*), parameter       :: F01  = "('(shr_strdata_set_stream_iodesc) ',a,i8,2x,i8,2x,a)"
-    character(*), parameter       :: F02  = "('(shr_strdata_set_stream_iodesc) ',a,i8,2x,i8,2x,i8,2x,a)"
+
+    integer                 :: stream_nlev
+    integer                 :: gsize2d
+    integer                 :: pio_iovartype
+    integer                 :: n, m, cnt
+    type(var_desc_t)        :: varid
+    integer                 :: ndims
+    character(CS)           :: dimname
+    integer, allocatable    :: dimids(:)
+    integer, allocatable    :: dimlens(:)
+    integer                 :: unlimdid
+    type(ESMF_DistGrid)     :: distGrid
+    integer                 :: lsize
+    integer, pointer        :: compdof(:) => null()
+    integer, pointer        :: compdof3d(:) => null()
+    integer                 :: rCode ! pio return code (only used when pio error handling is PIO_BCAST_ERROR)
+    character(*), parameter :: subname = '(shr_strdata_set_stream_iodesc) '
+    character(*), parameter :: F00  = "('(shr_strdata_set_stream_iodesc) ',a,i8,2x,i8,2x,a)"
+    character(*), parameter :: F01  = "('(shr_strdata_set_stream_iodesc) ',a,i8,2x,i8,2x,a)"
+    character(*), parameter :: F02  = "('(shr_strdata_set_stream_iodesc) ',a,i8,2x,i8,2x,i8,2x,a)"
     !-------------------------------------------------------------------------------
+
     rc = ESMF_SUCCESS
 
-!    call pio_seterrorhandling(pioid, PIO_BCAST_ERROR, old_error_handle)
+    ! set the number of vertical levels to a local variable
+    stream_nlev = per_stream%stream_nlev
```

Comment:
Why do we need a copy here?

Reply:
Because it makes it easier to read the code.